src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1409,68 +1481,102 @@ impl Cage {
                             "specified file not open for reading",
                         );
                     }
+                    // Check if the `O_NONBLOCK` flag is set in the pipe's flags.
+                    // If `O_NONBLOCK` is set, we set nonblocking to true, indicating that the pipe
+                    // operates in non-blocking mode.
                     let mut nonblocking = false;
                     if pipe_filedesc_obj.flags & O_NONBLOCK != 0 {
                         nonblocking = true;
                     }
+
+                    // Ensures that the function keeps trying to read data until it either succeeds
+                    // or encounters a non-retryable error.
                     loop {
                         // loop over pipe reads so we can periodically check for cancellation
                         let ret = pipe_filedesc_obj
                             .pipe
                             .read_from_pipe(buf, count, nonblocking)
                             as i32;
+                        // Check if the pipe is in blocking mode and the read returned EAGAIN.
+                        // It means that no data is currently available and the read should be tried
+                        // again later.
                         if pipe_filedesc_obj.flags & O_NONBLOCK == 0
                             && ret == -(Errno::EAGAIN as i32)
                         {
+                            // Check if the cancel status is set
                             if self
                                 .cancelstatus
                                 .load(interface::RustAtomicOrdering::Relaxed)
                             {
-                                // if the cancel status is set in the cage, we trap around a cancel
-                                // point until the individual thread
-                                // is signaled to cancel itself
+                                // If the cancel status is set in the cage, we trap around a cancel
+                                // point until the individual thread is signaled to cancel itself
                                 loop {
                                     interface::cancelpoint(self.cageid);
                                 }
                             }
-                            continue; //received EAGAIN on blocking pipe, try
-                                      // again
+                            // Received `EAGAIN` and no cancellation is requested, continue the loop
+                            // to try reading from the pipe again
+                            continue;
                         }
-                        return ret; // if we get here we can return
+                        // If the read was successful, return the result
+                        return ret;
                     }
                 }
+                // Reading from `Epoll` type file descriptors is not supported.
```

Comment:
I think for a match statement this shouldn't matter? I may be wrong.

Reply:
Yes, I think the same but its better to have things which fail together, so that the code looks more readable. Have updated it though in the latest commit.

Reply:
> Think of a match expression as being like a coin-sorting machine: coins slide down a track with variously sized holes along it, and each coin falls through the first hole it encounters that it fits into. In the same way, values go through each pattern in a match, and at the first pattern the value “fits,” the value falls into the associated code block to be used during execution.
- Rust book - [link](https://doc.rust-lang.org/book/ch06-02-match.html)

just mentioning here for clarity over the match workings